CSS color gets adjusted for disabled input elements

karlcow opened the pull request:
#### d9061430480df20b709c6a4014bf5eee587e8c73
<pre>
CSS color gets adjusted for disabled input elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=54643">https://bugs.webkit.org/show_bug.cgi?id=54643</a>
rdar://99826522

Reviewed by Tim Nguyen.

Currently WebKit overrides any style which is being chosen for
input:disabled by web developers. This fixes this by adjusting
a simpler default and making it possible for developers to adjust
the CSS. The solution proposed by Tim Nguyen is using color-mix().

* LayoutTests/platform/gtk/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/gtk/fast/forms/basic-textareas-expected.txt:
* LayoutTests/platform/gtk/fast/forms/basic-textareas-quirks-expected.txt:
* LayoutTests/platform/gtk/fast/forms/input-appearance-disabled-expected.txt:
* LayoutTests/platform/gtk/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/gtk/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/gtk/fast/forms/textarea-placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/ios-wk2/fast/forms/basic-textareas-quirks-expected.txt:
* LayoutTests/platform/ios/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/ios/fast/forms/basic-textareas-expected.txt:
* LayoutTests/platform/ios/fast/forms/input-appearance-disabled-expected.txt:
* LayoutTests/platform/ios/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/ios/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/ios/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/ios/fast/forms/textarea-placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/mac-bigsur/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-textareas-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-textareas-quirks-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-appearance-disabled-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/mac/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/mac/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/mac/fast/forms/textarea-placeholder-pseudo-style-expected.txt:
* Source/WebCore/css/html.css:
(input:disabled, textarea:disabled):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::CSSParserContext::CSSParserContext):
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::adjustInnerTextStyle const):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::disabledTextColor const): Deleted.
* Source/WebCore/rendering/RenderTheme.h:

Co-authored-by: Tim Nguyen &lt;ntim@apple.com&gt;

Canonical link: <a href="https://commits.webkit.org/260180@main">https://commits.webkit.org/260180@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/efd6fe5a8251e7b9d57681383a66774eac807340

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107450 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16511 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116612 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116033 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111348 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17935 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7752 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99613 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113211 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13538 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96717 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41161 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95443 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28343 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82932 "Found 1 new API test failure: /WebKitGTK/TestAuthentication:/webkit/Authentication/authentication-empty-realm (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9518 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29696 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10174 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6595 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15673 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49277 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7042 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11734 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/html/HTMLTextFormControlElement.cpp:
```diff
@@ -853,8 +853,9 @@ void HTMLTextFormControlElement::adjustInnerTextStyle(const RenderStyle& parentS
-        textBlockStyle.setColor(RenderTheme::singleton().disabledTextColor(textBlockStyle.visitedDependentColorWithColorFilter(CSSPropertyColor), parentStyle.visitedDependentColorWithColorFilter(CSSPropertyBackgroundColor)));
+    // FIXME: Remove me once we are sure, we get the right result.
+    // if (isDisabledFormControl())
+    //     textBlockStyle.setColor(textBlockStyle.visitedDependentColorWithColorFilter(CSSPropertyColor), parentStyle.visitedDependentColorWithColorFilter(CSSPropertyBackgroundColor));
```
It is not picking the color of html.css, I don't know why?

So this is not working yet. 
If I set a color in the document stylesheet as a webdev for `input:disabled` it is picking the right color.
BUT by default it doesn't seem to pick up the color specified by `color-mix` for the `textBlockStyle`. The border and widgets are taking the color. 

```
data:text/html,<!doctype%20html><html><input%20type="text"%20value="ABCDEFGHIJ"%20disabled="">
```


@weinig , Would you have a hint ?

mdubet replied:
I think `color-mix()` currently doesn't support system colors (such as `CanvasText`)

nt1m replied:
Works on latest trunk WebKit for me:

`data:text/html,<span%20style="color:color-mix(in%20hsl,%20CanvasText%2066%,%20Canvas);">fdsdfsdffddsf</span>`

karlcow replied:
yes the `color-mix()` is working when I set it manually, but it doesn't through `html.css` and I'm not sure why.

nt1m replied:
I found out why, https://searchfox.org/wubkat/rev/3de3ef30cefada3014150b0591598345dc7458e3/Source/WebCore/css/parser/CSSParserContext.cpp#57

You need to add `colorMixEnabled = true` there and it should work.

karlcow replied:
🤯 Thanks! Let's try.

karlcow committed:
CSS color gets adjusted for disabled input elements
https://bugs.webkit.org/show_bug.cgi?id=54643
rdar://99826522

Reviewed by Tim Nguyen.

Currently WebKit overrides any style which is being chosen for
input:disabled by web developers. This fixes this by adjusting
a simpler default and making it possible for developers to adjust
the CSS. The solution proposed by Tim Nguyen is using color-mix().

* LayoutTests/platform/gtk/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/gtk/fast/forms/basic-textareas-expected.txt:
* LayoutTests/platform/gtk/fast/forms/basic-textareas-quirks-expected.txt:
* LayoutTests/platform/gtk/fast/forms/input-appearance-disabled-expected.txt:
* LayoutTests/platform/gtk/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/gtk/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/gtk/fast/forms/textarea-placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/ios-wk2/fast/forms/basic-textareas-quirks-expected.txt:
* LayoutTests/platform/ios/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/ios/fast/forms/basic-textareas-expected.txt:
* LayoutTests/platform/ios/fast/forms/input-appearance-disabled-expected.txt:
* LayoutTests/platform/ios/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/ios/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/ios/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/ios/fast/forms/textarea-placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/mac-bigsur/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-inputs-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-textareas-expected.txt:
* LayoutTests/platform/mac/fast/forms/basic-textareas-quirks-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-appearance-disabled-expected.txt:
* LayoutTests/platform/mac/fast/forms/input-disabled-color-expected.txt:
* LayoutTests/platform/mac/fast/forms/number/number-appearance-spinbutton-disabled-readonly-expected.txt:
* LayoutTests/platform/mac/fast/forms/placeholder-pseudo-style-expected.txt:
* LayoutTests/platform/mac/fast/forms/textarea-placeholder-pseudo-style-expected.txt:
* Source/WebCore/css/html.css:
(input:disabled, textarea:disabled):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::CSSParserContext::CSSParserContext):
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::adjustInnerTextStyle const):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::disabledTextColor const): Deleted.
* Source/WebCore/rendering/RenderTheme.h:

Co-authored-by: Tim Nguyen <ntim@apple.com>

Canonical link: https://commits.webkit.org/260180@main

darinadler commented on Source/WebCore/css/html.css:
```diff
@@ -794,6 +794,10 @@ input::placeholder {
 }
 
+input:disabled, textarea:disabled {
+    color: color-mix(in hsl, CanvasText 66%, Canvas);
```
This new color is not dependent on the background color of the form element. So this means that in some cases when the background color is set, the old code would select a color that was visible and this new approach will allow the text to be invisible.

Perhaps that is not an important problem to solve?

Did you look at what the old and new approach do in various contexts aside from the ones where the color is set explicitly? I suppose if other web browsers don’t try to tackle this problem, we can just drop it. We are deciding that it’s up to the website designer to customize both the text and the background color.

nt1m replied:
> Did you look at what the old and new approach do in various contexts aside from the ones where the color is set explicitly? I suppose if other web browsers don’t try to tackle this problem, we can just drop it. We are deciding that it’s up to the website designer to customize both the text and the background color.

Other browsers just let the custom color also override the disabled color, there isn't any adjustment done.

karlcow replied:
This is the result in between different browsers and the patch. The situation is not all bad or all good. It depends on the case. So the previous algorithm was not also always a way to get a perfect contrasted rendering.

![Screenshot 2023-02-14 at 7 37 56](https://user-images.githubusercontent.com/505230/218590937-c3baf07e-e52f-40a2-b782-8563789cf8c9.png)

As @nt1m said the benefit here is that before designers were not able to control the color at all for the `disabled` mode, but now they can override the color to adjust to their choice colors matching their website design.

darinadler replied:
Yes, but that undeniable benefit — website authors can override the color — is mostly orthogonal to the thing I was questioning, the change to which color we choose when they **do not** override the color.

If our old choice of color was really good, we could have fixed this in a way that preserved our old algorithm while supporting overriding. The new color choice algorithm has to be judged on its merits.

I accept the screen shot as evidence that the new algorithm is better than the old. I am surprised that it somehow responds to the background. I didn’t realize how powerful `Canvas` and `CanvasText` were.
